src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -74,34 +112,45 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Solut
             // ⚠ DO NOT AWAIT INSIDE THE USING. The Dispose method that restores ExecutionContext flow must run on the
             // same thread where SuppressFlow was originally run.
             using var _ = FlowControlHelper.TrySuppressFlow();
-            return ReadDataSuppressedFlowAsync(pipeReader, solutionChecksum, objectCount, serializerService, callback, arg, cancellationToken);
+            return ReadDataSuppressedFlowAsync(pipeReader, objectCount, serializerService, callback, arg, cancellationToken);
 
             static async ValueTask ReadDataSuppressedFlowAsync(
-                PipeReader pipeReader, Checksum solutionChecksum, int objectCount, ISerializerService serializerService, Action<int, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
+                PipeReader pipeReader, int objectCount, ISerializerService serializerService, Action<Checksum, T, TArg> callback, TArg arg, CancellationToken cancellationToken)
             {
-                using var stream = await pipeReader.AsPrebufferedStreamAsync(cancellationToken).ConfigureAwait(false);
-                ReadData(stream, solutionChecksum, objectCount, serializerService, callback, arg, cancellationToken);
+                using var pipeReaderStream = pipeReader.AsStream(leaveOpen: true);
+
+                for (var i = 0; i < objectCount; i++)
+                {
+                    // First, read the length of the asset (and header) we'll be reading.
+                    var readResult = await pipeReader.ReadAtLeastAsync(sizeof(int), cancellationToken).ConfigureAwait(false);
+                    var length = ReadLength(readResult);
+
+                    // Advance past the length.
+                    pipeReader.AdvanceTo(readResult.Buffer.GetPosition(sizeof(int)));
+
+                    // Now buffer in the rest of the data we need to read.  Because we're reading as much data in as
+                    // we'll need to consume, all further reading (for this single item) can handle synchronously
+                    // without worrying about this blocking the reading thread on cross-process pipe io.
+                    await pipeReader.ReadAtLeastAsync(length, cancellationToken).ConfigureAwait(false);
+
+                    using var reader = ObjectReader.GetReader(pipeReaderStream, leaveOpen: true, cancellationToken);
```

Comment:
now we can just synchrnously read from that buffer which is now in memory in our process.